library/cool/gcd.red

```diff
@@ -0,0 +1,181 @@
+import prelude
+import cool.complete-induction
+import data.nat
+import data.void
+
+def double : nat → nat =
+  elim [
+  | zero → zero
+  | suc (n' → f) → suc (suc f)
+  ]
+
+def sub : nat → nat → nat =
+  elim [
+  | zero → λ _ → zero
+  | suc (m' → sub/m') →
+    elim [
+    | zero → suc m'
+    | suc n' → sub/m' n'
+    ]
+  ]
+
+def sub/le : (n m : nat) → le (sub n m) n =
+  elim [
+  | zero → λ _ → triv
+  | suc (n' → f) → λ m → elim m [
+    | zero → le/refl n'
+    | suc m' → le/suc/right (sub n' m') n' (f m')
+    ]
+  ]
+
+def mod/prop : nat → type = λ _ → nat → nat
+
+def mod : (n : nat) → mod/prop n =
+  let complete = weak/implies/complete mod/prop (λ P → realize/weak/induction P) in
+  complete
+    (λ _ → zero)
+    (λ n f → λ m →
+      elim m [
+      | zero → zero
+      | suc m' →
+        elim (sub (suc n) m) [
+        | zero →
+          elim (sub m (suc n)) [
+          | zero → zero
+          | suc _ → suc n
+          ]
+        | suc _ → (f (sub (suc n) (suc m')) (sub/le n m')) (suc m')
+        ]
+      ]
+    )
+
+def const/zero : nat → nat =
```

Comment:
this is a pretty confusing name!

Reply:
what about always/zero?

Reply:
isn't this the identity function, though?

Reply:
oops you're right. eta-expansion?

Reply:
maybe id/nat? the next definition is already named eta...

Reply:
I was gonna change it to eta-equiv, but that works (commuting conversion is the right word probably)